desktop/src/gui/theme.rs

```diff
@@ -0,0 +1,209 @@
+    async fn start_dbus_theme_watcher_linux(&self) {
+        async fn start_inner(this: &ThemeController) -> Result<(), Box<dyn Error>> {
+            let data = this.data();
+            let Some(ref connection) = data.zbus_connection else {
+                return Ok(());
+            };
+
+            let settings = FreedesktopSettings::new(connection).await?;
+
+            let mut stream = Box::pin(settings.watch_color_scheme().await?);
+
+            let this2 = this.clone();
+            tokio::spawn(Box::pin(async move {
+                while let Some(scheme) = stream.next().await {
+                    match scheme {
+                        Ok(scheme) => {
+                            this2.set_theme(scheme_to_theme(scheme));
+                        }
+                        Err(err) => {
+                            tracing::warn!(
+                                "Error while watching for color scheme changes: {}",
+                                err
+                            );
+                        }
+                    }
+                }
+            }));
+
+            Ok(())
+        }
+
+        if let Err(err) = start_inner(self).await {
+            tracing::warn!("Error registering theme watcher: {}", err);
+        }
+    }
+
+    fn data(&self) -> MutexGuard<'_, ThemeControllerData> {
+        self.0.try_lock().expect("")
+    }
+
+    fn set_theme_internal(&self, data: MutexGuard<'_, ThemeControllerData>, theme: Theme) {
+        data.egui_ctx.set_visuals(match theme {
+            Theme::Light => egui::Visuals::light(),
+            Theme::Dark => egui::Visuals::dark(),
+        });
+        if let Some(window) = data.window.upgrade() {
+            window.request_redraw();
+        }
+    }
+
+    pub fn set_theme(&self, theme: Theme) {
+        let data = self.data();
+        if data.theme_preference != ThemePreference::System {
+            // Cannot change theme when there's a preference.
+            return;
+        }
+        self.set_theme_internal(data, theme);
+    }
+
+    async fn set_theme_preference(&self, theme_preference: ThemePreference) {
+        let mut data = self.data();
+        data.theme_preference = theme_preference;
+        let theme = match theme_preference {
+            ThemePreference::System => {
+                if let Ok(theme) = self
+                    .get_system_theme()
+                    .await
+                    .inspect_err(|err| tracing::warn!("Unable to read system theme: {err}"))
+                {
+                    theme
+                } else {
+                    return;
+                }
+            }
+            ThemePreference::Light => Theme::Light,
+            ThemePreference::Dark => Theme::Dark,
+        };
+        self.set_theme_internal(data, theme);
+    }
+
+    #[cfg(target_os = "linux")]
+    async fn get_system_theme(&self) -> Result<Theme, Box<dyn Error>> {
+        let connection = zbus::Connection::session().await?;
```

Comment:
Can you not reuse the open connection in self?

Reply:
Whoops, of course, that was the main reason of the refactor, just... forgot about that :smile: thanks!